File: windows/ClipSync.Core/Media/MediaLimits.cs
namespace ClipSync.Core.Media;

/// <summary>Stage 9 static clipboard image limits. Shared by storage, protocol v2, and capture.</summary>
public static class MediaLimits
{
    public const int MaxEncodedBytes = 16 * 1024 * 1024;
    public const int MaxPixels = 32 * 1024 * 1024;
    public const int MaxSide = 8_192;
    public const int MaxChunkBytes = 256 * 1024;
    public const int MaxConcurrentDownloads = 2;
    public const int UnfinishedDownloadHours = 24;
    public const int ThumbnailMaxSide = 512;
    public const int MaxImagePayloadBatchEncodedBytes = MaxEncodedBytes;
    public const int MaxChunkCount = 64;

    public const string MimePng = "image/png";
    public const string MimeJpeg = "image/jpeg";
    public const string KindImage = "image";
    public const string KindText = "text";

    public const string BlobStateReady = "ready";
    public const string BlobStatePending = "pending";
    public const string BlobStateFailed = "failed";

    public const string ClipMediaReady = "ready";
    public const string ClipMediaPending = "pending";
    public const string ClipMediaMissing = "missing";

    public static bool IsSupportedMime(string? mime) =>
        string.Equals(mime, MimePng, StringComparison.Ordinal)
        || string.Equals(mime, MimeJpeg, StringComparison.Ordinal);

    public static bool FitsPixelBudget(int width, int height)
    {
        if (width is < 1 or > MaxSide || height is < 1 or > MaxSide)
        {
            return false;
        }

        return (long)width * height <= MaxPixels;
    }
}

File: android/app/src/main/java/com/clipsync/android/media/MediaLimits.kt
package com.clipsync.android.media

/** Stage 9 static clipboard image limits. Shared by storage, protocol v2, and capture. */
object MediaLimits {
    const val MAX_ENCODED_BYTES = 16 * 1024 * 1024
    const val MAX_PIXELS = 32 * 1024 * 1024
    const val MAX_SIDE = 8_192
    const val MAX_CHUNK_BYTES = 256 * 1024
    const val MAX_CONCURRENT_DOWNLOADS = 2
    const val UNFINISHED_DOWNLOAD_HOURS = 24
    const val BLOB_GC_GRACE_MS = 5L * 60L * 1000L
    const val THUMBNAIL_MAX_SIDE = 512
    const val MAX_CHUNK_COUNT = 64

    const val MIME_PNG = "image/png"
    const val MIME_JPEG = "image/jpeg"
    const val KIND_IMAGE = "image"
    const val KIND_TEXT = "text"

    const val BLOB_STATE_READY = "ready"
    const val BLOB_STATE_PENDING = "pending"
    const val BLOB_STATE_FAILED = "failed"

    const val CLIP_MEDIA_READY = "ready"
    const val CLIP_MEDIA_PENDING = "pending"
    const val CLIP_MEDIA_MISSING = "missing"

    fun isSupportedMime(mime: String?): Boolean =
        mime == MIME_PNG || mime == MIME_JPEG

    fun fitsPixelBudget(width: Int, height: Int): Boolean {
        if (width !in 1..MAX_SIDE || height !in 1..MAX_SIDE) {
            return false
        }
        return width.toLong() * height.toLong() <= MAX_PIXELS
    }
}
